src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionTextCache.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq.Expressions;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    /// <summary>
+    /// This class holds the cache for the expression text that is computed by ExpressionHelper.
+    /// Expressions of node type member, constant and parameter are cached.
+    /// However expressions that include indexers are not cached.
+    /// </summary>
+    public class ExpressionTextCache
+    {
+        public ConcurrentDictionary<LambdaExpression, string> Entries { get; } = 
+            new ConcurrentDictionary<LambdaExpression, string>(LambdaExpressionComparer.Instance);
+
+        private class LambdaExpressionComparer : IEqualityComparer<LambdaExpression>
+        {
+            public static readonly LambdaExpressionComparer Instance = new LambdaExpressionComparer();
+
+            public bool Equals(LambdaExpression lambdaExpression1, LambdaExpression lambdaExpression2)
+            {
+                // We will cache only pure member access expressions. Hence we compare two expressions 
+                // to be equal only if they are identical member access expressions.
+                var expression1 = lambdaExpression1.Body;
+                var expression2 = lambdaExpression2.Body;
+                var identical = false;
+
+                while (true)
```

Comment:
Return `true` if `ReferenceEquals(expression1, expression2)` before diving into checking the details.